ARROW-8918: [C++][Python] Implement cast metafunction to allow use of "cast" with CallFunction, use in Python

wesm opened the pull request:
This provides the `CAST(data AS target_type)` SQL idiom. The target_type is provided via CastOptions (FWIW I believe this is the most correct approach for handling the target_type). As a result we no longer need to maintain separate binding boilerplate in Python for Array vs. ChunkedArray

jorisvandenbossche commented on python/pyarrow/_compute.pyx:
```diff
@@ -316,3 +322,58 @@ cdef class FilterOptions(FunctionOptions):
+    <pyarrow.lib.TimestampArray object at 0x10420eb88>
+    [
+      1262304000000,
+      1420070400000
```
Minor note, but for me (on master), this gives actual string date representation instead of milliseconds:

```
In [6]: arr.cast(pa.timestamp('ms'))
Out[6]: 
<pyarrow.lib.TimestampArray object at 0x7f95c16bcc48>
[
  2010-01-01 00:00:00.000,
  2015-01-01 00:00:00.000
]
```

jorisvandenbossche replied:
Ah, I see now it is just copied from the old docstrings below. That explains why the old repr is used here ;)

jorisvandenbossche commented on python/pyarrow/_compute.pyx:
```diff
@@ -316,3 +322,58 @@ cdef class FilterOptions(FunctionOptions):
+    """
+    Cast array values to another data type.
+
+    Example
+    -------
```
```suggestion
    Examples
    --------
```

(typically the examples are also put after Parameters / Returns sections)

wesm commented:
I had copy-pasted the original docstring for `Array.cast`. I'll clean it up per above

wesm commented:
Fixed. I also moved the cast Python wrapper to compute.py so it doesn't need to be compiled

wesm commented:
I missed that there was a partial wrapper for CastOptions used in _dataset.pyx, fixing that now

wesm commented:
I just moved the compute wrappers to compute.pxi (so they are part of `pyarrow.lib` now) and fixed up the usage of CastOptions between `pyarrow.compute.cast` and `_dataset.pyx`. So this patch will need to be merged before doing more work on the compute bindings

wesm commented:
Could someone review this?

jorisvandenbossche commented:
Is there a specific reason for moving compute into `pyarrow.lib` as well? (I would think multiple cython compilation units might be more "scalable" regarding code maintenance etc)

jorisvandenbossche commented:
For the rest looks good to me!

wesm commented:
@jorisvandenbossche I can try moving it back, I was concerned about having to `cimport` things back into lib.pyx

jorisvandenbossche commented:
What would need to be cimported in lib.pyx? I think in the Array and Table classes etc, for the computational methods, you do the `_pc.call_function(..)`. Or is it for the Option classes? (`_dataset.pyx` could also cimport that from `_compute`, like it does for the FileSystem for example)

(now, I don't know how important this is, it was just a throught)

wesm commented:
I added a _compute.pxd, no big deal. Will merge this once the build passes

wesm commented:
Merging. The ARM failure (not sure what went wrong?) does not need to block
